### DISCUSSION TITLE
Violations where Resource Undefined

### DISCUSSION THREAD

**Discussion by pranav-super** [NASA-AMMOS/aerie#248] [Ideas]
This turns out not to be as simple as our discussion on Slack might have implied.

For context, the ticket that this pertains to is one that resolves [AERIE-1825]( https://jira.jpl.nasa.gov/projects/AERIE/issues/AERIE-1825?filter=allopenissues). This ticket involves doing constraints checking on external datasets (which I will also refer to as external resources or external profiles, as that is the language we adopted when working on this ticket). Currently, it is possible to upload external profiles, and this ticket takes the profiles that have been uploaded, integrates them with the constraints DSL, and makes it so that violations are checked on these resources. That is, if I have an external profile “/ext1” uploaded, the constraint `export default () => Constraint: {Real.Resource(“/ext1”).greaterThan(20)}` becomes valid.

Generally, this is a pretty straightforward ticket, except for the fact that it is totally possible that these external resources have sparse profile segments. That is, they may not be defined everywhere. If we call `+` a space where the resource is defined and `-` a space where it isn’t, then it is totally possible that we could have the following profile:
 
> /plant: --333--22--
(undefined from `t=[1,2],[6,7],[10,11]`)

That part isn’t the problem – implementing that hasn’t been much of an issue, and it can address some useful test cases where a resource is based on something like telemetry data, or pertains to a component that isn’t used or attached to the mission for the entirety of a mission, like an alternative power source.

The problem is constraints checking itself. That is, should we flag a constraint as being violated when the resource the constraint operates on isn’t defined? So should the constraint `Real.Resource(“/plant”).equal(3.0)` yield a violation everywhere except `t=[3,5]`, or only in `t=[8,9]`?

After some deliberation on Slack, where the two conflicting opinions were:
-	if it isn’t defined, then the constraint can’t be true, so return a violation
-	if it isn’t defined, telling the user their constraint was violated is misleading and shouldn’t be done

the latter of the two was chosen. That is, **when a resource was undefined do not to return a violation**. That’s about where slack leaves off. There was discussion of ways to implement this – we want to do this in a way where we don’t really mess with existing implementations of methods like `evaluate()` because the scheduling DSL uses the exact same Java tree nodes, and we don’t want a change in constraints checking to mess with scheduling. This is because currently, the default is a comparison with undefined returns false, but what we imply here is a comparison with undefined should return true. The solution was to isolate all constraint-specific changes the the one constraint-specific class, `ViolationsOf.java`, and then do any windowing restrictions using a special method that we add to the `Expression.java` class, `getWhereDefined()`, which returns an intersection (this will be qualified shortly, it’s not always an intersection) of where the leaf nodes’ resources are defined, and then takes the windows that were found from normal constraints checking and intersects it with the result of that call to `getWhereDefined()`. This was a cool solution, except for the fact that it isn’t really well defined for nodes like `ForEachActivity`, and there are two cases where it really just doesn’t make a ton of sense, and this has to do with composite constraints:
1.	Let’s say we have an `Any` constraint, where one of the subconstraints is defined but false in a given window, and the other subconstraint’s resource is undefined in that same window. In that case, constraints checking evaluates that composite constraint as false, as undefined is treated as false. But should it return a violation? On one hand, none of them are true, so it’s clearly not satisfied, but on the other hand is a constraint isn’t satisfied is the composite constraint completely defined? We leaned heavily towards the latter, but the second case was what really broke this solution.
2.	Let’s say we have an `All` constraint, with one subconstraint true in a given window, and the other undefined in the same window. This would evaluate to false, as undefined equals false, and if one is false, all will be. But should we really return a violation? On one hand, __technically__ not every constraint is satisfied, so that’s grounds for violation. But that’s not totally consistent with the idea that an undefined constraint should cause a violation – the constraints that __are__ defined are true, so this shouldn’t be a violation. This is definitely up for debate, but the vagueness here led to Joel proposing another solution, which we also want comment on.

The solution was to revert to the initial behavior, i.e. undefined = violation, **unless explicitly stated otherwise**. That is, we add a new node to the DSL, called `undefined()`, as a method on `Real.Resource()`, and it returns the windows for which the resource isn’t defined. Then, by doing the following:
 
`Real.Resource("hello").whenDefined().lessThan(3);`
(apologies, I can't get multiline code to work here)

you’ve said this is true when the resource is undefined, or where it is defined and it is less than 3, in other words, only evaluate this constraint when this resource is defined.

But, imaginably, users would need to type this often. There could be cases where they might want constraints checking to fail when a resource is undefined (we think?), so this verbosity in some form is important as we would then want users to have the option to not type this (this could be up for debate), but maybe there’s a more concise way. The solution was to add a method to `Real.Resource()` called `.whenDefined()`, that translates as follows:

- user types: `Real.Resource("hello").whenDefined().lessThan(3);`
- `.whenDefined()` returns `NullableReal.Resource("hello")`, where `NullableReal.Resource` is the same as `Real.Resource` except every time you call something like `.lessThan()` or `.equals()`, it returns the expanded form above, i.e. it returns:
- the final result, or `Real.Resource("hello").whenDefined().lessThan(3);`

But there might be a cleaner easier way to do this, so comment on that too would be appreciated as well.

After we figure these details out, though, we can finish off [AERIE-1825]( https://jira.jpl.nasa.gov/projects/AERIE/issues/AERIE-1825?filter=allopenissues)!

**Comment by Twisol** [comment]
> if it isn’t defined, then the constraint can’t be true,

If we're missing a span of telemetry, it doesn't mean the resource didn't take on a desired value; it means we don't *know* what value it took on. It very well may have taken on the desired value, but we can't observe that.

> if it isn’t defined, telling the user their constraint was violated is misleading and shouldn’t be done

This is my opinion, too. If data is missing, we can say neither that a condition is met nor that it is not met -- we simply don't have the data to make such a determination.

Classically, this kind of use-case is what the `NULL` value in databases is for: missing data. NULL compares weirdly with itself: almost any operation involving NULL produces NULL.

<details>
<summary>Example <code>psql</code> session</summary>

Because of this, almost any computation involving a NULL produces NULL itself.

```
postgres=# \pset null '<null>'
Null display is "<null>".

postgres=# \pset format csv
Output format is csv.

postgres=# select null + 1 as result;
result
<null>

postgres=# select null + 1 as result;
result
<null>

postgres=# select null > 1 as result;
result
<null>

postgres=# select null = null as result;
result
<null>
```
</details>

> This is because currently, the default is a comparison with undefined returns false, but what we imply here is a comparison with undefined should return true.

I would argue that a comparison with `undefined` should return *`undefined`*, assuming `undefined` plays the same role as NULL above.

It kind of seems to me that a condition needs to not only produce windows on which it is true, but also windows on which it is undefined (or windows on which it is false; pick any two, they imply the third). This works pointwise, just like anything else with conditions: a condition operator evaluated at time t is defined if its operands are also defined at time t (or whatever other time it consults, like with translating or scaling operators).

> Let’s say we have an Any constraint

As long as one of Any's subconditions is true at a time, the Any should be true at that time, regardless of the value of the other subconditions. This is because, regardless of whether the other subconditions resolve to true or false, it *doesn't change the answer* that the `Any` would be true either way.

Conversely, for `Any` to be false at a time, all of its subconditions must be `false`. A single `null` amongst otherwise all-`false`s will cause the `Any` to be undefined at that time.

> Let’s say we have an All constraint

As long as one of All's subconditions is false at a time, the All should be false at that time, regardless of the value of the other subconditions. Conversely, for `All` to be true at a time, all of its subconditions must be `true`.

**Reply by Twisol** [reply]
In general, the principle here is that, if the answer *would not change* regardless of what value the `null` input was replaced with, then you're allowed to give a non-`null` output. If you could get different answers depending on the resolution of the `null`, you must produce a `null`.

**Reply by Twisol** [reply]
I do want to point out that this principle isn't fully observed in databases like Postgres. For instance, if you specify  `WHERE x = x`, and `x` happens to be `NULL`, the clause is considered `false` even though it's impossible for any concrete resolution of `x` to falsify `x = x`. So they selectively break this principle for various reasons.

<details>
<summary>Example `psql` session</summary>

```
postgres=# create table lol (x int null);
CREATE TABLE

postgres=# insert into lol values (null);
INSERT 0 1

postgres=# insert into lol values (1);
INSERT 0 1

postgres=# select x as result from lol where x = x;
result
1

postgres=# select x = x as result from lol;
result
<null>
t
```

(It's perfectly legal to produce `NULL` even when you *could have* given a better answer, so the last result doesn't break this principle. You might imagine that in more complicated cases, proving that `t` is an allowable response might be very costly.)
</details>

**Reply by pranav-super** [reply]
But in context of constraints-checking, what would this look like? Should I return two sets of windows, one where there definitively are constraints violations, and one where there are nulls that obscure the answer? How should that be presented to the user?

> I would argue that a comparison with undefined should return undefined, assuming undefined plays the same role as NULL above.

I think @JoelCourtney has some good examples of where this exactly happens and what I mean by that, I don't think I conveyed his sentiment correctly here (we had a discussion earlier, prompting this post).

**Reply by dyst5422** [reply]
Quick point - we should use null instead of undefined. Undefined can be returned unexpectedly by indexing operations etc. (using a not yet defined variable or indexing into an array at a location it isn't defined). Null is not commonly returned by any built-ins and should be the designator for such non-existent values to avoid uncertainty.

**Reply by pranav-super** [reply]
`null` is what is returned and passed around, the only use of `undefined()` was just the name of an EDSL method.

**Reply by JoelCourtney** [reply]
I like @Twisol's solution here. The whole reason for all of this is because currently nulls have to be discarded and "resolved" somehow before being transformed into Windows, because Windows can't support it. Pranav, the solutions we talked about on Thursday are ways for either explicitly or implicitly getting rid of nulls. If we instead change Windows and Violations to be nullable too, then we wouldn't have this problem.

The only issue I can think of with that is scheduling. If a windows-type argument for (i.e.) `analyseWhen` is null, what should the scheduler do?

**Reply by pranav-super** [reply]
But for something like constraints, it seems like we either want a true or a false, right? It seems like it might be a little confusing for users if we have two different highlights in the UI (or something like that) indicating there are null issues and also legitimate constraints violations.

And when you say nullable window, what exactly does that mean? Updating window so it can handle operations on null?

**Reply by JoelCourtney** [reply]
I'm sure we all already know this, but in this context, "null" isn't actually a value that is explicitly `null`, its a gap.

Currently windows either exist or don't. *Mechanically* this is very similar to how profiles work. At any given time, a profile is either a value from its [codomain](https://en.wikipedia.org/wiki/Codomain), or nothing (null). Windows are represented like a special kind of profile on the codomain `{true}`. They are either `true` or null.

*Practically* speaking though, we are already using null in Windows to mean `false`, not truly null. What Jonathan is proposing is that we add expand the codomain to `{true, false}`, so that null can be used to *actually* mean null. This would allow comparison with null to return a null that _actually means_ null instead of `false`. This would also remove the problems we had for `Any` and `All`:

Input | Any | All
---|---|---
T,N | T | N
F,N | N | F
T,F,N | T | F

Then we could do two things that would make this really understandable for the user:
1. Have the UI mark nulls in the top-level constraint as some kind of warning
2. Allow the author to transform nulls in profiles and windows with a method like `.assignGaps(value)`, where `value` is something in the codomain.

**Reply by Twisol** [reply]
Thanks for the cogent summary, @JoelCourtney. :+1:

I like the `assignGaps` operator. This lets users decide how to interpret gaps, while being very explicit about the decision and the context in which it's being made.

> _Mechanically_ this is very similar to how profiles work. [...] Windows are represented like a special kind of profile on the codomain `{true}`. They are either `true` or null.

Yes (almost). As you say, you can think of a resource profile as a (piecewise-defined) function from time to some codomain. Conditions are (piecewise) functions from time to *the booleans*; at any given time, the condition is either true *or false*. It just so happens that a function into the booleans is the same as a set: if `f(t) == true` then `t` is in the set. We leverage that to represent windows in a convenient/compact way, by simply recording the regions where it's true. We call the intervals of time on which a condition is true its "windows", and we can recover the regions where it's false automatically from its windows.

When we bring missing information into the mix, profiles and conditions both become *partial* functions -- they may not have a value. That might make the current representation for windows less tenable; the "false" set can't be inferred just from the "true" set anymore. But windows and profiles *are* structured in the same way, even if they're not necessarily represented that way in the current system, so insights from one can absolutely inform the other.

It seems halfway plausible, to me, that we could represent conditions much like a discrete resource: as a series of intervals on which the condition takes on some value. Null / missing information at a time would be represented the same way as for a resource: as a lack of an interval covering that time.

**Reply by Twisol** [reply]
Incidentally, I suspect that having a well-grounded and consistent perspective on missing information will also feed into the results streaming work @skovati is working on. While a simulation is in progress, any results not yet returned *are precisely missing information*.

**Reply by pranav-super** [reply]
The discrete resource idea I think makes this make a lot more sense, even though I think its saying what has been stated earlier. I'll see what I can figure out based on this.

**Reply by Twisol** [reply]
I'm also poking around to see if the `IntervalSet` class we have in Java can be made into an `IntervalMap` instead. (Recalling that a `Set<K>` is little more than a `Map<K, Boolean>`.) Don't gate anything on me though, I'm just exploring.

**Reply by pranav-super** [reply]
I was talking to Joel and he was suggesting removing any dependence of Windows on `IntervalSet`s, and not to touch `IntervalSet` or `IntervalAlgebra` at all. I do not know about how this works so I can't speak to any of it or what should change though, and I'm not sure what removing dependence of Windows on `IntervalSet`s means beyond changing line 15 of `Windows.java` to use a different object.

**Reply by Twisol** [reply]
Yeah, it would be good to hear what @JoelCourtney has in mind -- Windows is little more than a thin wrapper around IntervalSet with a chosen IntervalAlgebra, so as far as I can imagine, divorcing the two would amount to reimplementing the functionality -- not dissimilar to implementing an IntervalMap, I think?

**Reply by pranav-super** [reply]
Would it be worth implementing an `IntervalMap` interface? Would that be used by anything else other than `Windows`?

**Reply by JoelCourtney** [reply]
I'd trust Jonathan's advice on how to implement this more than mine, cuz he implemented it in the first place :). I do know that I'll want `IntervalAlgebra` to remain the same, but I don't think that's on the table.

**Reply by Twisol** [reply]
> Would it be worth implementing an `IntervalMap` interface? Would that be used by anything else other than `Windows`?

I don't remember if anything much else even uses IntervalSet, either. But I find it easier to think about how these containers should work if I keep them generic.

> I do know that I'll want `IntervalAlgebra` to remain the same, but I don't think that's on the table.

It's only a hunch for now, but I think IntervalAlgebra won't need to change.

**Reply by pranav-super** [reply]
I'm partway through implementing `IntervalMap<I,V>` where `I` is some sort of interval and `V` is any sort of value, but you mentioned earlier that you were poking around with this - did you end up finding anything or have any recommendations that you think are worth considering going into this implementation?

**Reply by Twisol** [reply]
I'll post my code when I get back home, but I have the following thoughts:

* Operations like intersect() need to be moved to `Windows`, since they're specialized to combining boolean values between two collections. On the other hand, `add` becomes `set`, and it becomes a little more complicated because we have to meld or snip any intervals overlapping with the given one depending on whether the associated values are equal.
* `subtract` between two general maps seems kind of odd. At least the way I imagine a general `subtract`, we wouldn't care about the values in the subtrahend, just the intervals. So I think it's a little ill-posed. We can very easily have `unset` (aka `remove`) on one interval at a time, though, which is really what `subtract` was getting at.
* Ideally we'd have `map()` and `map2()` methods for performing unary and binary operations on `IntervalMap`s; then I think most operations of interest boil down to these two, which generically provide the pattern of traversal currently employed by `intersect` and `subtract`.
* A map with the same implementation strategy as our current set will take ~double the storage, because we're explicitly storing each interval that used to be implied by the gaps. In non-sparse maps where all times have a value on some interval, the start of one interval is the same as the end of the one prior (with a flipped inclusivity). We could switch to a totally different representation where each interval is represented by its start time, and its end time is implied by the start of the next interval. (This is *exactly* the representation used for resources in the database.) Unfortunately, I think this would require moving away from `IntervalAlgebra` entirely.

**Reply by pranav-super** [reply]
> Operations like intersect() need to be moved to `Windows`...

I did what you suggested with `intersect()` and `add` and all in `IntervalMap`, where I would split overlapping intervals based on value; if we have it in `Windows` it might make `IntervalMap` just a thin and unnecessary wrapper over `Map` with little added functionality. That being said, having it in `Windows` allows us to explicitly take advantage of the value type for the map being boolean, if necessary (I didn't have to thus far in implementing `IntervalMap` but presumably it could be of use).

> `subtract` between two general maps seems kind of odd

What I did was just make it between the map and a parameter `Iterable<Interval>`, and just remove everything from the map that the parameter overlapped with.

> because we're explicitly storing each interval that used to be implied by the gaps

To clarify, this is because we would store true where the interval exists (i.e. where a constraint is satisfied), false where it doesn't (i.e. where a constraint is not satisfied), and then nothing/null only if the interval is legitimately undefined (i.e. an external constraint doesn't exist), right?

> We could switch to a totally different representation where each interval is represented by its start time, and its end time is implied by the start of the next interval.

At that point, we would need our map to have order to it as well, or we would need to sort the intervals associated with the map every time we did an operation, right?

**Reply by Twisol** [reply]
> if we have it in `Windows` it might make `IntervalMap` just a thin and unnecessary wrapper over `Map` with little added functionality.

I'm just not sure what `intersect({[1, 3]: "foo", [2, 4]: "bar"})` ought to be. Clearly it's `{[2, 3]: _}`, but what value is the `_`? It seems like the answer must depend on the type we're working with, and thus a generic `IntervalMap<I, V>` can't know what to do without more context.

> What I did was just make it between the map and a parameter `Iterable<Interval>`, and just remove everything from the map that the parameter overlapped with.

Yes, so essentially a loop over an `unset` with a single `Interval` parameter? One of the benefits of the existing `IntervalSet` is that it can efficiently subtract two sets, because we only need to iterate once over the set being subtracted from.

Also, it's important for `subtract` that we *keep* the part of the interval that doesn't overlap; [1, 3] - [2, 4] is [1, 2). I think this is what you meant, but we don't wholly remove intervals that merely overlap.

> To clarify, this is because [...]

Right.

> At that point, we would need our map to have order to it as well

The `IntervalSet` is already ordered, particularly for traversal; the `IntervalMap` should accommodate this anyway.

> it might make `IntervalMap` just a thin and unnecessary wrapper over `Map` with little added functionality

My prototype of `IntervalMap` still uses an ordered list in its implementation; just, instead of `List<I>`, it's `List<Pair<I, V>>`, directly associating an explicit value to each interval. Java's own `Map` is good for discrete queries, where you have an exact match between the stored key and the key to query on, but temporal queries (particularly "find containing interval") are a poor fit.

The `ArrayList` used by the current `IntervalSet` is already not ideal for long timelines with many intervals, as we do lots of in-place `remove`s and `add`s that require shuffling all elements along. We should come back and tune things later, just as Postgres does (its tables are just lists, and indices are extra structures that organize those records for faster access). But for now, I think we're mostly just changing what the intervals are associated with (from implicit booleans to explicit arbitrary values), rather than changing the overall behavior of the structure.

**Reply by Twisol** [reply]
Here's what I wrote so far; I implemented `set` and `unset`, as well as `map`. I want to implement `map2` and see if it can cleanly capture the patterns of traversal you see in `set` and `unset` (and in the original `IntervalSet`); if that's the case, it would be nice to only have to document and explain that traversal in one place.

https://gist.github.com/Twisol/206a8144d24635df17520eed0a86954f

I paused this to see what it would take to implement the alternative representation I mentioned that's more space-efficient. I don't know what that will look like, yet.

**Reply by pranav-super** [reply]
What is `map2` supposed to do?

Also, I went through the code for `IntervalMap` and using a `List` instead of a map is a far better and easier solution than what I had. What's the difference between `subtract` and `unset`? And going back to your question:

> I'm just not sure what `intersect({[1, 3]: "foo", [2, 4]: "bar"})` ought to be. Clearly it's `{[2, 3]: _}`, but what value is the `_`?

I don't totally see how we might want a `Windows`-specific implementation here. That value is definitely vague - I was thinking if `IntervalMap A = {[1,3]: "foo"}` and we call intersect with the parameter `IntervalMap B = {[2,4]: "bar"}`, where the call is `A.intersect(B)`, we would take the intersection and adopt the value of the parameter (`bar`). This is seemingly arbitrary, but if we switch to the `Windows`-specific context, where the value is either only `true` or `false`, which value would we prefer, and why?

**Reply by Twisol** [reply]
> What is `map2` supposed to do?

It's easiest to explain by analogy with `Optional`.

With `Optional<A>`, we have [a method](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/Optional.html#map(java.util.function.Function)) `<B> Optional<A> map(Function<A, B> transform)`. (The real type has extra variance bounds.) We can write that Haskell-style as `(A -> B) -> Optional A -> Optional B`, where I've made `this` the last parameter. In words, if we have an Optional A and a way to transform an A into a B, we can produce an Optional B. "Naturally", if given optional is empty, the result is also empty (we can't conjure a B from nowhere); likewise, if it's not empty, then the result is an Optional containing the return value of the transformation applied to the input value.

Lots of types can support a `map` operation, including `List` and our own `EventGraph`. But `map` is only good for unary transformations; we can build up pipelines that transform some initial input, but the pipeline only ever takes one input. `map2` lets us combine pipelines: we take *two* input values and a function that combines them, and produce *one* output value.

Haskell-wise, this looks like `(A -> B -> C) -> Optional A -> Optional B -> Optional C` -- given a transform and two inputs optionals, we compute an output optional using that transform.

So `map2` on `IntervalMap` is the same idea. We've given an `IntervalMap I A` and an `IntervalMap I B`, and we want to compute an `IntervalMap I C`. That is, if time `t` has value `a` on one timeline, and value `b` on another timeline, then time `t` on the result is some function of the two values at that time.

**Reply by Twisol** [reply]
> but if we switch to the `Windows`-specific context, where the value is either only `true` or `false`, which value would we prefer, and why?

The `intersect` and `union` operations on sets of `V` are equivalent to logical `AND` and `OR` operations on functions `V -> Bool`. So, if at some time we have `true` on one side and `false` on the other, `intersect` will take `false` and `union` will take `true`.

**Reply by pranav-super** [reply]
> That is, if time t has value a on one timeline, and value b on another timeline, then time t on the result is some function of the two values at that time.

That makes sense.

I think I'll switch to your list-based `IntervalMap` implementation as that seems significantly more straightforward, but outside of `IntervalMap`, for methods like `intersect`, within a `Windows`-specific context, I'm not totally sure what to do. You suggested we adopt Boolean logic of `AND`s and `OR`s, which I think makes sense (instead of obfuscating the value entirely, which would be the case if it was implemented in `IntervalMap`), but should we then also handle the `null` case here, or is there a better place for that?.

Also, what is the design reasoning behind having `IntervalMap` in the first place, if `Windows` is all that uses it? Would this scale to allow for something else `Windows`-like that maps to another values, and what could that even be?

**Reply by Twisol** [reply]
> Also, what is the design reasoning behind having `IntervalMap` in the first place, if `Windows` is all that uses it? Would this scale to allow for something else `Windows`-like that maps to another values, and what could that even be?

I'd like to see `IntervalMap` used for resource profiles, too. Like I noted a bit earlier, the alternate representation I'm pondering is exactly what we're already using for profiles in the database. We know both profiles and conditions need to cope with missing information, and I've always struggled to identify in what way they're fundamentally *different* structures. So I think there's a very strong chance that `IntervalMap` will be used as the foundation for both things.

**Reply by pranav-super** [reply]
> I'd like to see IntervalMap used for resource profiles, too.

As an alternative to what we have for `DiscreteProfile` and `DiscreteProfilePiece`, for example?

**Reply by Twisol** [reply]
> but should we then also handle the `null` case here, or is there a better place for that?.

Yeah, we'll have to. For instance, `intersect(wins1, wins2)` can ultimately be implemented as `map2(wins1, wins2, (a$, b$) -> a$.flatMap(a -> b$.map(b -> a && b)))`, where the transformation taken by `map2` is an `Optional A -> Optional B -> Optional C`, not just an `A -> B -> C`.

**Reply by Twisol** [reply]
> As an alternative to what we have for `DiscreteProfile` and `DiscreteProfilePiece`, for example?

Yes. Probably `DiscreteProfile` would become like `Windows` in that it would internally use an `IntervalMap` to store the data currently governed by its `Piece`s.

**Reply by pranav-super** [reply]
So after I adopt your version of `IntervalMap`, what all should be done following that? It seems the next steps are:

- make sure everything in `Windows` is using the right operation in `IntervalMap` where it applies (i.e. `subtract` operations use `unset`)
- figure out `map2` in `IntervalMap`
- implement operations like `intersect`, which are now wholly in `Windows`
- maybe update the resource profiles?

I just want to be clear on what the next steps are so I'm not doing work you're already doing or have already done.

Also, I'm not sure how feasible this is, but would you be available to meet sometime tomorrow to discuss some of this?

**Reply by Twisol** [reply]
> I just want to be clear on what the next steps are so I'm not doing work you're already doing or have already done.

For the time being, you can think of me as just a scout; I'm checking the path ahead to see what options and obstacles might exist. I might implement `map2` to satisfy my own concerns about what that looks like -- the algorithmic logic in `IntervalSet` has always been rather hairy -- but I don't expect to go beyond that. Arguably, it'd be a *good* thing for both of us to implement it. We can run test cases against both to hammer out any bugs and inconsistencies. There's no doubt in my mind that `map2` is the most technically challenging part; but its advantage is that we ought to be able to write most of the other key methods in terms of it.

Then, as long as we settle on an API for `IntervalMap` that doesn't depend on its internal representation, the alternate representation I'm pondering can always be tackled later on.

> maybe update the resource profiles?

You'll have to see @patkenneally or Basak for legitimate marching orders, but I'd argue it's beyond the scope of the current discussion. We definitely want to design in that direction, but I think the immediate and pressing concern is handling missing information in conditions. Unless there are issues to tackle on the resource profile side as well, we should plan to clean up and unify our foundations only after we've addressed the immediate problem.

(I like to draw on multiple problems and concerns to come up with flexible designs, but one problem still gets priority here, I think.)

> Also, I'm not sure how feasible this is, but would you be available to meet sometime tomorrow to discuss some of this?

🤔 I should be free. My email address is in my profile :+1: I prefer afternoon times if at all possible. Just keep in mind that I'm not employed by JPL anymore, so I can't speak authoritatively for the project direction.

**Reply by Twisol** [reply]
> What's the difference between `subtract` and `unset`? And going back to your question:

I don't think I ever answered this. The idea with `unset` is that it makes the timeline undefined along the given interval. With `subtract`, we were making our timeline false everywhere the given timeline was true.

We could define `unsetAll` using `map2` now: given `wins1 : IntervalMap I A` and `wins2 : IntervalMap I Unit` (where "Unit" is the enum with only one value), we apply `map2(wins1, wins2, (a$, b$) -> b$.isPresent() ? Optional.empty() : a$))`. The fact that we're using `Unit` captures the fact that we only care about presence or absence. This is the closest to what we mean by "subtract" in a general setting.

> For instance, `intersect(wins1, wins2)` can ultimately be implemented as

I think `intersect` needs the below as a transform; the one I gave before is, I think incorrect.

```java
  (a$, b$) -> {
    if (a$.isPresent()) {
      return (a$.get()) ? b$ : a$;
    } else {
      return (b$.orElse(true)) ? a$ : b$;
    }
  }
```

Still easier to verify this than to convince oneself that the current `intersect` is correct, I will argue.

**Reply by pranav-super** [reply]
> Arguably, it'd be a good thing for both of us to implement it...as long as we settle on an API for IntervalMap that doesn't depend on its internal representation, the alternate representation I'm pondering can always be tackled later on.

Okay. So for now, as I implement some of this, you're thinking I should keep the API you suggested for `IntervalMap`, it's just a question of how `map2` should be implemented and later `Windows` from that.

> (I like to draw on multiple problems and concerns to come up with flexible designs, but one problem still gets priority here, I think.)

Sounds good!

> My email address is in my profile  

I sent you a meeting invite.

**Reply by Twisol** [reply]
> Okay. So for now, as I implement some of this, you're thinking I should keep the API you suggested for `IntervalMap`, it's just a question of how `map2` should be implemented and later `Windows` from that.

Yes, basically. But the API I suggested is just what I came up with; if issues come up we should re-evaluate it under our design constraints.

**Reply by pranav-super** [reply]
> For instance, `intersect(wins1, wins2)` can ultimately be implemented as

When you give this anonymous function, do you mean this is what is passed as an argument to `map` as the transform, and said call to `map` would be the whole implementation of `intersect`?

**Reply by Twisol** [reply]
To `map2`, but yes, exactly. The anonymous function I gave fills the hole in `map2(wins1, wins2, _)`.

**Reply by pranav-super** [reply]
This is what I got for `map2`; I'm [linking it](https://gist.github.com/pranav-super/79bb41080515744001a6c9e7d56e4df1) for reference when we meet later. There are a lot of cases (4) that need to be handled, which led to a lot of slightly varied repetition.

**Reply by pranav-super** [reply]
@Twisol Suggestion from Joel - change IntervalMap<Alg, I, V> where I is the abstract interval to just IntervalMap<V>. The only allowed interval should just be Window, and the only allowed algebra then, is WindowsAlgebra. The reasoning here is that there is really no other way to represent a Window. There are alternatives, such as a Window which has start and end time, versus start time and duration, but there will come points where we have to compare these or work with both - it would be ideal if we just have one Window type to go with and when they are represented differently, simply convert. So if we pull from the database and windows are stored differently than what Window dictates, we would convert it at that stage and then store it.

At that point, as Window would be the monolithic interval type, and therefore WindowsAlgebra is the only IntervalAlgebra type, we could merge them as just Window and WindowAlgebra, dropping the interface and the genericness entirely. 

After doing that we could rename Window to Interval just because we don't want to marry the Window class by name to the Windows class, as Spans will use that Window class as well.

**Reply by Twisol** [reply]
> @Twisol Suggestion from Joel - change IntervalMap<Alg, I, V> where I is the abstract interval to just IntervalMap.

To `IntervalMap<V>`, keeping the `V` generic, I assume?

Yeah, I think I agree. I'd still keep the `WindowAlgebra` implementation separate, as I do *like* defining those trait operations separately from `IntervalMap` for compartmentalization's sake, but I don't see any alternate implementations of `IntervalAlgebra` materializing anytime soon, so you may as well instantiate `alg` within `IntervalMap` itself.

**Reply by Twisol** [reply]
You could also just move all the methods on the `IntervalAlgebra` interface to `Window` I guess. `Window` is kind of messy, but I guess it's ultimately the right place for those methods to live.

**Reply by pranav-super** [reply]
I was thinking everything should be merged.

**Reply by Twisol** [reply]
Probably sensible.

**Reply by Twisol** [reply]
I haven't tested it yet *at all*... but here's my tentative `map2`? https://gist.github.com/Twisol/9bb074d5721f63b97c7b7ecbcbbb37b6

(This is now my *fifth* implementation strategy. I mentioned I was only up to four during our chat, lol.)

**Reply by Twisol** [reply]
I've just edited the above `map2` to make things slightly less baroque. I shouldn't be making any further edits.

**Reply by Twisol** [reply]
I lied -- I noticed a bug in how the interval remainders are computed that would cause problems if `overlap` was empty. (The handled prefix would not get removed from the interval, so we would, in all likelihood, enter an infinite loop.) I fixed it in the gist.

**Reply by Twisol** [reply]
Think I spotted another bug, an off-by-one error this time. Fixed.

I'm still poking at the alternate representation -- I'm writing `map2` for it now, so I'm still reviewing my earlier `map2` and adapting it to the setting.

**Reply by Twisol** [reply]
I've just "finished" my alternate representation for `IntervalMap`. Again, I haven't tested this at all -- it compiles, and that's all I know for sure. https://gist.github.com/Twisol/88a8b2bf9e71681e74c1b4160ee9a03e

The `IntervalMap` API still works with intervals, but this implementation only stores the individual transitions between profile states. There's an iterator for running over all defined intervals, and an append-only builder for creating new instances. I ended up making `IntervalMap` itself immutable for simplicity, I think we generally want fast forward traversal and whole-timeline transformations, not random read-write access.

I also included three general-purpose methods that use `map2`, namely `mask` (implementing the generalized form of @JoelCourtney's `applyWhen`), `select` (pare the timeline down to only intervals intersecting a given window), and `zip` (generically pair values between two maps). These demonstrate how `map2` abstracts over all the mess of traversing two maps, only asking you to specify what value the result should have based on the values (or absence thereof) on each input map.

(The `hull()` method gives the bounding interval containing all segments; this would support the `minTimePoint` and `maxTimePoint` methods on `Windows`.)

**Reply by pranav-super** [reply]
...I did not see this.

I also have an implementation of `IntervalMap` that I just finished yesterday - I also found an issue in `map2` that had to do with when both input maps in `map2` were null between intervals, but I fixed that and implemented a new version of Windows on top of it.

This was based off of the first implementation of `map2` you sent, by the way.

But I think it's definitely important that we factor in your `IntervalMap` implementation, I just missed this update :(

**Reply by Twisol** [reply]
It would be good to test the behavior of one against the other, to see if we understood the problem constraints the same way 😄

**Reply by pranav-super** [reply]
Fortunately I wrote some pretty extensive map2 tests, but then again it was based on how I was approaching the problem

**Comment by JoelCourtney** [comment]
Hi, I'm the annoying guy who keeps changing the conversation to talk about Spans.

So what happens when a user converts a `Windows` that has gaps into `Spans`? My first instinct is that `null` makes negative sense in the context of `Spans`, because Spans is just "a set of intervals", nothing more. But then converting Windows to Spans becomes a lossy operation, because `null` will be lost when it becomes "not an interval".

So the question is, does it make any sense to include the concept of a null span, so that the conversion is lossless? If so, how would it behave under operations like `split`, `shorterThan`, or `starts`? My vote is to officially define them as follows:

- **Windows:** a boolean profile, which can be null because profiles can be null
- **Spans:** a set of intervals, which can't be null because what even the heck

Any objections?

**Reply by Twisol** [reply]
This interpretation fits with my mental model of the topic, so I'm 100% on board 😄
